subsys/bluetooth/controller/ll_sw/ull_adv_sync.c

```diff
@@ -161,8 +161,115 @@ uint8_t ll_adv_sync_param_set(uint8_t handle, uint16_t interval, uint16_t flags)
 uint8_t ll_adv_sync_ad_data_set(uint8_t handle, uint8_t op, uint8_t len,
 				uint8_t const *const data)
 {
-	/* TODO */
-	return BT_HCI_ERR_CMD_DISALLOWED;
+	struct pdu_adv_com_ext_adv *ter_com_hdr, *ter_com_hdr_prev;
+	struct pdu_adv_hdr *ter_hdr, ter_hdr_prev;
+	struct pdu_adv *ter_pdu, *ter_pdu_prev;
+	uint8_t ter_len, ter_len_prev;
+	uint8_t *ter_dptr, *ter_dptr_prev;
+	struct lll_adv_sync *lll_sync;
+	struct ll_adv_set *adv;
+	uint8_t ter_idx;
+
+	adv =  ull_adv_is_created_get(handle);
+	if (!adv) {
+		return BT_HCI_ERR_UNKNOWN_ADV_IDENTIFIER;
+	}
+
+	lll_sync = adv->lll.sync;
+	if (!lll_sync) {
+		return BT_HCI_ERR_UNKNOWN_ADV_IDENTIFIER;
+	}
+
+	/* Get reference to previous tertiary PDU data */
+	ter_pdu_prev = lll_adv_sync_data_peek(lll_sync);
+	ter_com_hdr_prev = (void *)&ter_pdu_prev->adv_ext_ind;
+	ter_hdr = (void *)ter_com_hdr_prev->ext_hdr_adi_adv_data;
+	ter_hdr_prev = *ter_hdr;
+	ter_dptr_prev = (uint8_t *)ter_hdr + sizeof(*ter_hdr);
+
+	/* Get reference to new tertiary PDU data buffer */
+	ter_pdu = lll_adv_sync_data_alloc(lll_sync, &ter_idx);
+	ter_pdu->type = ter_pdu_prev->type;
+	ter_pdu->rfu = 0U;
+	ter_pdu->chan_sel = 0U;
+	ter_com_hdr = (void *)&ter_pdu->adv_ext_ind;
+	ter_com_hdr->adv_mode = ter_com_hdr_prev->adv_mode;
+	ter_hdr = (void *)ter_com_hdr->ext_hdr_adi_adv_data;
+	ter_dptr = (uint8_t *)ter_hdr + sizeof(*ter_hdr);
+	*(uint8_t *)ter_hdr = 0U;
+
+	/* No AdvA */
+	/* No TargetA */
+
+	/* TODO: CTEInfo */
+
+	/* No ADI */
+
+	/* TODO: AuxPtr */
+
+	/* No SyncInfo */
+
+	/* TODO: Tx Power flag */
+
+	/* TODO: ACAD */
+
+	/* AdvData */
+	/* Calc previous tertiary PDU len */
+	ter_len_prev = ull_adv_aux_hdr_len_get(ter_com_hdr_prev, ter_dptr_prev);
+
+	/* Did we parse beyond PDU length? */
+	if (ter_len_prev > ter_pdu_prev->len) {
+		/* we should not encounter invalid length */
+		return BT_HCI_ERR_UNSPECIFIED;
+	}
+
+	/* Calc current tertiary PDU len */
+	ter_len = ull_adv_aux_hdr_len_get(ter_com_hdr, ter_dptr);
+	ull_adv_aux_hdr_len_fill(ter_com_hdr, ter_len);
+
+	/* Set the tertiary PDU len */
+	ter_pdu->len = ter_len + len;
+
+	/* Check AdvData overflow */
+	if (ter_pdu->len > CONFIG_BT_CTLR_ADV_DATA_LEN_MAX) {
+		/* FIXME: release allocations */
+		return BT_HCI_ERR_PACKET_TOO_LONG;
+	}
+
+	/* Start filling tertiary PDU payload based on flags from here
+	 * ==============================================================
+	 */
+
+	/* Fill AdvData in tertiary PDU */
+	memcpy(ter_dptr, data, len);
+
+	/* TODO: Fill ACAD in tertiary PDU */
+
+	/* Tx Power */
+	if (ter_hdr->tx_pwr) {
+		*--ter_dptr = *--ter_dptr_prev;
```

Comment:
It looks kinda odd to me, that you decrement the data pointer to get the last octet in the header. To me, it would make a lot more sense to something like 
```
ter_hdr[TX_POWER_OFFSET] = ter_hdr_prev[TX_POWER_OFFSET]
```

Unless I'm misreading this line?

Reply:
Or even better, since the struct has the field, why not do

```
ter_hdr->tx_pwr = ter_hdr_prev->tx_pwr;
```

Reply:
The values are present based on bitmap in the 8-bit header that precedes the values. Hence, fixed offset value is not applicable here. Structure member de-referencing cannot be used.

Reply:
> Structure member de-referencing cannot be used.

Why not?

Reply:
![image](https://user-images.githubusercontent.com/6350656/98185957-06432900-1f0e-11eb-81b5-c595653a4e99.png)

> If a flag bit is set to 1, the corresponding Extended Header field is present;
otherwise, the corresponding Extended Header field is not present.

bit fields in Extended Header Flags define the presence of rest of the values, hence a fixed offset will not work.

Reply:
Ah, it took a while, but I think I understand now. 

But I think you might have an off-by-one error here: As far as I can tell, you assign `ter_dptr` as
```
ter_dptr = (uint8_t *)ter_hdr + sizeof(*ter_hdr);
```

but do not increment the `ter_dptr` pointer any time before this statement. When you do `*--ter_dptr`, don't you overwrite the extended header flags, rather than writing to the octet that should contain the TxPower? It looks like you are missing an octet between the extended header flags, and the ACAD when you copy the AD data in line 244.

Reply:
> But I think you might have an off-by-one error here: As far as I can tell, you assign `ter_dptr` as
> 
> ```
> ter_dptr = (uint8_t *)ter_hdr + sizeof(*ter_hdr);
> ```
> 

`ter_dtpr` points to the byte after `ter_hdr`, it is not off-by-one.

>When you do `*--ter_dptr`, don't you overwrite the extended header flags, rather than writing to the octet that should contain the TxPower?

No, ter_dptr has been incremented in line in the `TODO` at line 212 ;-)

Reply:
Ah :) 

I think it might be worth either adding a note here, or commenting the block until the `TODO` in 212 have been done then, as that certainly isn't clear from the current state